### PR TITLE
Simple lego in Rust

### PR DESCRIPTION
**Pull request by kvinwang**
Due the the poor performance of the lego written in JS, let use this one to support the hackthon.

# Note
**please upgrade cargo-contract to 3.2 to compile this PR.**

### CODE REVIEW

File: phat/contracts/lego_rs/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "pink/std",
+    "scale/std",
+    "scale-info/std",
+    "pink-json/std",
+    "serde/std",
+]
+ink-as-dependency = []
+e2e-tests = []
+
+[patch.crates-io]
```
Comment (shelvenzhou) [inline phat/contracts/lego_rs/Cargo.toml:37]: Do we need patches here to support big JS input?

File: phat/contracts/lego_rs/lib.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        #[ink(message)]
+        pub fn version(&self) -> VersionTuple {
+            version_tuple!()
+        }
+
+        #[ink(message)]
+        pub fn run(&self, actions: String) -> bool {
+            info!("lego_rs: actions={}", actions);
```
Comment (shelvenzhou) [inline phat/contracts/lego_rs/lib.rs:62]: We shall not log actions here. E.g. in the future we may transfer some sensitive values among actions.

File: phat/contracts/lego_rs/lib.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                return false;
+            };
+            pipeline(&actions)
+        }
+    }
+
+    fn pipeline(actions: &[Action]) -> bool {
+        let mut result = String::new();
+        for action in actions {
+            info!("running action={:?}", action);
```
Comment (shelvenzhou) [inline phat/contracts/lego_rs/lib.rs:74]: Same as above.

File: phat/contracts/lego_rs/lib.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            match action {
+                Action::Call { callee, selector } => {
+                    let res = invoke_contract(From::from(*callee), 0, 0, *selector, &[], false);
```
Comment (shelvenzhou) [inline phat/contracts/lego_rs/lib.rs:77]: How to pass the output from previous action to the latter one?

Comment (shelvenzhou) [reply phat/contracts/lego_rs/lib.rs:77]: Also it's okey to forbid reentry here, but we should record this in doc.

File: phat/contracts/lego_rs/Cargo.toml (hunk 1 above)
Comment (kvinwang) [reply phat/contracts/lego_rs/Cargo.toml:37]: This PR is only a workaround for the performance issue of the JS lego.It just implements the only used command `call` without args in the bricks. So, looks like it doesn't need big IO buffer.

File: phat/contracts/lego_rs/lib.rs (hunk 2 above)
Comment (kvinwang) [reply phat/contracts/lego_rs/lib.rs:62]: Aagin, this PR is for special usage. When we do the full implmementation, the log would be controled by the `debug` flag in the workfow manifest in the future. As it defined in #3.

File: phat/contracts/lego_rs/lib.rs (hunk 4 above)
Comment (kvinwang) [reply phat/contracts/lego_rs/lib.rs:77]: > How to pass the output from previous action to the latter one?

No such feature at the moment.

Comment (kvinwang) [reply phat/contracts/lego_rs/lib.rs:77]: > Also it's okey to forbid reentry here, but we should record this in doc.

A full implementation will let the caller config the reentrant.

File: phat/contracts/lego_rs/lib.rs (hunk 2 above)
Comment (shelvenzhou) [reply phat/contracts/lego_rs/lib.rs:62]: Created an issue #21 to record this.

**Review by shelvenzhou — APPROVED** [review]

**Review by h4x3rotab — APPROVED** [review]

**Pull request merged**